game/src/main/kotlin/world/gregs/voidps/world/activity/skill/herblore/HerbCleaning.kts

```diff
@@ -0,0 +1,40 @@
+package world.gregs.voidps.world.activity.skill.herblore
+
+import world.gregs.voidps.engine.data.definition.data.Cleaning
+import world.gregs.voidps.engine.entity.character.player.Player
+import world.gregs.voidps.engine.entity.character.player.skill.Skill
+import world.gregs.voidps.engine.entity.character.player.skill.level.Level.has
+import world.gregs.voidps.engine.entity.item.Item
+import world.gregs.voidps.engine.inv.inventory
+import world.gregs.voidps.engine.inv.replace
+import world.gregs.voidps.engine.queue.weakQueue
+import world.gregs.voidps.world.interact.entity.player.equip.inventoryOption
+
+
+/*
+Click each herb to clean faster
+Click 1 herb and auto clean with a 3 tick delay
+ */
+inventoryOption("Clean", "inventory") {
+    val herb: Cleaning = item.def.getOrNull("cleaning") ?: return@inventoryOption
+    if (!player.has(Skill.Herblore, herb.level, true)) {
+        return@inventoryOption
+    }
+    cleanHerb(slot, player, item, herb)
+    player.clean(item, player, herb)
+}
+
+fun Player.clean(item: Item, player: Player, herb: Cleaning) {
+    if(!player.inventory.contains(item.id)) {
+        return
+    }
+    weakQueue("cleaning", 3) {
+        cleanHerb(player.inventory.indexOf(item.id), player, item, herb)
+        clean(item, player, herb)
+    }
```

Comment:
This doesn't seem correct, if you click one herb with more than one in the inventory it'll clean 2 of them?

I don't think there's a queue involved here as to my knowledge herblore can't be cancelled? Unless there's some mechanic I'm forgetting I think it just uses the skilling clock to prevent cleaning from happening too frequently

Reply:
> This doesn't seem correct, if you click one herb with more than one in the inventory it'll clean 2 of them?
No it cleans one i will attach a GIF

[
![cleaning](https://github.com/GregHib/void/assets/79747812/c6e2b890-6304-44fb-a4b9-d3138d2da594)
](url)

> I don't think there's a queue involved here as to my knowledge herblore can't be cancelled? Unless there's some mechanic I'm forgetting I think it just uses the skilling clock to prevent cleaning from happening too frequently

if i understand right what you're asking; in RuneScape when cleaning herbs you can click a single herb and it will automatically clean them with no interaction needed but you can speed cleaning up buy click each herb in your inventory,  I am not not quite sure what you mean by it can't be cancelled? 

Video here: https://youtu.be/waww7COFxbU?si=yqk0TxLozG7R4o_t&t=15

Reply:
Cancelled meaning you can interrupt and prevent the action from occuring part way through, like walking away when cooking or fishing. That's the use-case for queue's anyway.

Yeah so in that clip he's cleaning more than 1 herb per tick, so I don't think there's anything that needs doing other than just:

```kotlin
inventoryOption("Clean", "inventory") {
    val herb: Cleaning = item.def.getOrNull("cleaning") ?: return@inventoryOption
    if (!player.has(Skill.Herblore, herb.level, true)) {
        return@inventoryOption
    }
    
    if (player.inventory.replace(slot, item.id, item.id.replace("grimy", "clean"))) {
        player.experience.add(Skill.Herblore, herb.xp)
    }
}
````

Reply:
Okay no problem, The reason why i made it the way I did was because i was trying to reproduce this behaviour where it would clean them automatically after x ticks unless the player clicks the herb to clean them that's why i used the queue.  for example in osrs:

![osrsexample](https://github.com/GregHib/void/assets/79747812/89b2c60c-8114-49e0-9aba-77af67dae5f0)

Reply:
Does it not work the same when you try in-game? I would've thought it automatically occurs from the way packets are stored and processed at the start of each tick

Reply:
Just so we're both on the same page, because i think theres some confusion, when spam clicking the herb in Void it works as expected like OSRS, But in OSRS when i click the first herb i leave my mouse theres a timer that will clean the next herb in 2 ticks till all herbs are cleaned. if i were to replace the code i have with what you suggested it would removed the behaviour where if i leave my mouse it would clean the next herb in 2 ticks till they're all cleaned, sorry if i am not being clear.

Reply:
Okay so the first herb has a 2 tick delay and all ones after that are instant or also have 2 tick delay? You're right about the softQueue for the 2 ticks, what I didn't understand is why it was calling cleanHerb before the queue as well

Reply:
Okay so when you click the grimy herb it will clean that herb instantly then after that every 2 ticks it will clean the next herb that's why i call `cleanHerb(slot, player, item, herb)` to handle the instant cleaning then i call `player.clean(item, player, herb)` to handle the 2 tick delay cleaning

Reply:
That's bizarre, that's not how I remember it working in rs2, osrs wiki has some mention of a herblore update but I can't find the exact mention of changes to the mechanics.

Reply:
You're right there were changes in September 20th 2020 is when they actually added the automatically cleaning mechanic in the update post is says "Clean Herbs; You may now clean all your herbs automatically after cleaning the first one in your Inventory." I completely missed this, and i always thought it was in RS2, would you like me to remove the automatically cleaning mechanic or keep it for QOL?

Reply:
I think let's leave it out for now, I've added it to #58 as it is a nice qol improvement but it could also be done with a more standard way with the make/clean x interface. I haven't really decided how closely to emulate vs enhance the gameplay so I've been making notes and putting stuff under the "beyond emulation" issue tag as things to do in the future once all the core content is complete

Reply:
No problem i have removed it, sorry about the confusion, I wasn't aware that was actually added 3 years ago, i thought it was a legacy mechanic

Reply:
No problem, I didn't even realise it was a mechanic in osrs, show's how often I play